spring-integration-core/src/main/java/org/springframework/integration/handler/LoggingHandler.java

```diff
@@ -164,58 +169,66 @@ protected void onInit() throws Exception {
 	@Override
 	protected void handleMessageInternal(Message<?> message) throws Exception {
 		switch (this.level) {
-		case FATAL:
-			if (this.messageLogger.isFatalEnabled()) {
-				this.messageLogger.fatal(createLogMessage(message));
-			}
-			break;
-		case ERROR:
-			if (this.messageLogger.isErrorEnabled()) {
-				this.messageLogger.error(createLogMessage(message));
-			}
-			break;
-		case WARN:
-			if (this.messageLogger.isWarnEnabled()) {
-				this.messageLogger.warn(createLogMessage(message));
-			}
-			break;
-		case INFO:
-			if (this.messageLogger.isInfoEnabled()) {
-				this.messageLogger.info(createLogMessage(message));
-			}
-			break;
-		case DEBUG:
-			if (this.messageLogger.isDebugEnabled()) {
-				this.messageLogger.debug(createLogMessage(message));
-			}
-			break;
-		case TRACE:
-			if (this.messageLogger.isTraceEnabled()) {
-				this.messageLogger.trace(createLogMessage(message));
-			}
-			break;
-		default:
-			throw new IllegalStateException("Level '" + this.level + "' is not supported");
+			case FATAL:
+				if (this.messageLogger.isFatalEnabled()) {
+					this.messageLogger.fatal(createLogMessage(message));
+				}
+				break;
+			case ERROR:
+				if (this.messageLogger.isErrorEnabled()) {
+					this.messageLogger.error(createLogMessage(message));
+				}
+				break;
+			case WARN:
+				if (this.messageLogger.isWarnEnabled()) {
+					this.messageLogger.warn(createLogMessage(message));
+				}
+				break;
+			case INFO:
+				if (this.messageLogger.isInfoEnabled()) {
+					this.messageLogger.info(createLogMessage(message));
+				}
+				break;
+			case DEBUG:
+				if (this.messageLogger.isDebugEnabled()) {
+					this.messageLogger.debug(createLogMessage(message));
+				}
+				break;
+			case TRACE:
+				if (this.messageLogger.isTraceEnabled()) {
+					this.messageLogger.trace(createLogMessage(message));
+				}
+				break;
+			default:
+				throw new IllegalStateException("Level '" + this.level + "' is not supported");
 		}
 	}
 
-
 	private Object createLogMessage(Message<?> message) {
+		Assert.notNull(this.expression, "'expression' must not be null");
```

Comment:
No, it's not funny.
The `onInit()` must be there anyway. And its purpose to provide `BeanFactoryAware` `expression` and `evaluationContext`. 
I'm not sure which side effects might happen if you use the `LoggingHandler` without `BeanFactory`, but as you insist we should reinstate the backward compatibility for this components. As Gary says: cat is out of bag already. We can only deprecate such a wrong behavior.
But seems for me we have spent already so much time for so simple problem.

If I would do a new component I advocate just only for the  `BeanFactoryAware` behavior, but since we have a case when some code is already used it improperly and there is no so easy way to change that code so, we decided to meet your requirements and reinstate an old behavior.
That's all. I don't see any other arguments to discuss further.